### PR TITLE
Optimise dvc ls -R

### PR DESCRIPTION
**Pull request by rlamy**
More optimisations for `dvc ls --recursive`. According to my informal local benchmark, it brings its run time from 38s to 20s.

### CODE REVIEW

File: dvc/fs/dvc.py [hunk 1]
```diff
@@ -292,7 +297,22 @@ def _get_fs_pair(
         Returns a pair of fss based on repo the path falls in, using prefix.
         """
         key = self._get_key_from_relative(path)
+        repo, fs, fs_path, dvc_fs, dvc_parts = self._get_fs_pair_2(key)
+        dvc_path = _get_dvc_path(dvc_fs, dvc_parts)
+        return repo, fs, fs_path, dvc_fs, dvc_path
 
+    def _get_fs_pair_2(
```
Comment (efiop) [inline dvc/fs/dvc.py:304]: The naming is kinda bad 😄

File: dvc/fs/dvc.py [hunk 2]
```diff
@@ -79,6 +79,11 @@ def _merge_info(repo, fs_info, dvc_info):
     return ret
 
 
+def _get_dvc_path(dvc_fs, dvc_parts):
+    if dvc_fs:
```
Comment (efiop) [inline dvc/fs/dvc.py:83]: Seems like in both uses we already check that dvc_fs is not none

Comment (rlamy) [reply dvc/fs/dvc.py:83]: `_get_fs_pair()` doesn't check (indeed, I moved the check from there) but it seems that `dvc_fs` is never None anyway.

File: dvc/fs/dvc.py (hunk 1 above)
Comment (rlamy) [reply dvc/fs/dvc.py:304]: Yeah, I know! I was hoping someone would have a better idea...

Comment (efiop) [reply dvc/fs/dvc.py:304]: @rlamy The `get_fs_pair` is also not great, it just stuck around from old times while changing the functionality. We should probably settle the naming for both. Your fresh look should be able to find better names for both (or feel free to reorganize those functions as you please to make new stuff make more sense).

Comment (rlamy) [reply dvc/fs/dvc.py:304]: Yes, the naming issue is a sign that `_get_fs_pair_2()` doesn't make much sense...

File: dvc/fs/dvc.py [hunk 3]
```diff
@@ -109,11 +94,6 @@ def __init__(
 
         from pygtrie import Trie
 
-        if repo is None:
-            repo, repo_factory = self._repo_from_fs_config(
-                subrepos=subrepos, **kwargs
-            )
-
```
Comment (efiop) [inline dvc/fs/dvc.py:116]: This is dead, but let's keep it as it is useful and is actually needed for making dvcfs public.

**Pull request merged**